Add Client Certificates tutorial. (#1148)
mendersoftware/mender-docs: 08.Server-integration/03.Client-certificates/docs.md

---
title: Client certificates
taxonomy:
    category: docs
---

> Mender supports setting up an ambassador server which can authenticate devices using TLS client certificates. Each client is equipped with a certificate signed by a CA certificate (Certificate Authority), and the ambassador authenticates devices by verifying this signature. Devices that are authenticated are automatically authorized in the Mender backend, and do not need to be manually approved.

Reviewer: `ambassador server` sounds strange; we call it ambassador because that's the name of the pattern in distributed architectures, but I'm not 100% sure this is understandable by people outside of the project. The technical name is edge proxy, which stands for "reverse proxy at the edge of (your) network".

This is in particular useful in a mass production setting because you can sign client certificates when they are manufactured so they automatically get accepted into the Mender server when your customer turns them on, which might happen several months after manufacturing.

See [Device authentication](../../02.Overview/13.Device-authentication/docs.md) for a general overview of how device authentication works in Mender.

## Prerequisites

### A board integrated with Mender

You need a physical board that has already been integrated with Mender. For example, you may use one of the reference boards BeagleBone Black, Raspberry Pi 3 or Raspberry Pi 4.

If you have not yet prepared a device visit one of the following:

- [Client installation](../../03.Client-installation/chapter.md)
- [System updates: Debian family](../../04.System-updates-Debian-family/chapter.md)
- [System updates: Yocto Project](../../05.System-updates-Yocto-Project/chapter.md)

### Mender client and server connectivity

<!-- TODO, this section may need to be rewritten slightly. It's not really incorrect, but it doesn't fit perfectly either, since we will set up the ambassador. -->

> Once your device boots with a newly provisioned disk image, it should already be correctly connecting to the Mender server. After booting the device you see it pending authorization in the Mender server UI, similar to the following.

Author: I think this doesn't carry the same meaning. In fact, I would go in the opposite direction, and change

* After booting the device you see it pending authorization"

-into-

* After booting the device you **should** see it pending authorization

The important point here is that these are things you need to check, because there are no preceding instructions, so we can't automatically expect it to be true.

> Once your device boots with a newly provisioned disk image, it should already be correctly connecting to the Mender server. After booting the device you see it pending authorization in the Mender server UI, similar to the following.

Reviewer 2: Makes sense :)

![Mender UI - device pending authorization](device-pending-authorization.png)

If your device does not show as pending authorization in the Mender server once it is booted with the disk image, you need to diagnose this issue before continuing. See the [troubleshooting section on connecting devices](../../201.Troubleshoot/05.Device-Runtime/docs.md#mender-server-connection-issues) in this case.

### A CLI environment for your server

> In order to access the the Mender server API with the commands below, you need to set up some shell variables in the terminal you will be using.

Author: Yep, both shorter and better!

Follow the steps in [set up shell variables for cURL](../01.Using-the-apis/docs.md#set-up-shell-variables-for-curl).

### Mender-Artifact tool

Download the `mender-artifact` tool from the [Downloads section](../../09.Downloads/docs.md).

## Generate certificates

In this section we will generate and sign certificates for the server and the devices.

### Generate a CA certificate

<!--AUTOVERSION: "generate a % certificate"/ignore-->
First we need to generate a master certificate which will be used to sign each client certificate. We start by generating a private key:

```bash
openssl ecparam -genkey -name P-256 -noout -out ca-private.key
```

!!! The "P-256" curve can be switched with a different curve if necessary.

Next we create a configuration file which contains information about the Certificate Authority. Execute the following command to create the file:

```bash
cat > ca-cert.conf <<EOF
[req]
distinguished_name = req_distinguished_name
prompt = no

[req_distinguished_name]
commonName=my-server.com
organizationName=My Organization
organizationalUnitName=My Unit
emailAddress=myusername@example.com
countryName=NO
localityName=Oslo
stateOrProvinceName=Oslo
EOF
```

> The fields should be filled with information about your organization, locality and contact information. In particular, make sure `commonName` matches the domain name of the server which will serve as the mTLS ambassador. <!-- TODO, link to ambassador setup -->

Reviewer: reword:
```matches the domain name of the edge proxy which will serve as the mTLS ambassador.```

Then generate a certificate from the newly generated private key:

```bash
openssl req -new -x509 -key ca-private.key -out ca-cert.pem -config ca-cert.conf -days $((365*10))
```

! The `-days` argument specifies how long the certificate is valid, and can be adjusted if needed. The example expression gives a certificate which is valid for approximately 10 years. Since the CA certificate will only be used on the Mender server, it is usually not important that it expires, and it's better to have a long expiry time to avoid having to rotate certificates on the devices.

### Generate a client certificate

> When preparing client certificate for a device, the certificate key will be generated on a separate system (not on the device), and then provisioned into the device storage. This way we can keep records of the public key of the device and ensure sufficient entropy during key generation, so the resulting keys are secure random.

Author: In this case, I think we should use "can" or some equivalent, because we are not instructing the user, just vaguely mentioning something they can and possibly should do.

So I think "can keep" is better, but is using "we" bad? Should that be written differently?

> When preparing client certificate for a device, the certificate key will be generated on a separate system (not on the device), and then provisioned into the device storage. This way we can keep records of the public key of the device and ensure sufficient entropy during key generation, so the resulting keys are secure random.

Reviewer 2: Not necessarily. [This](https://github.com/mendersoftware/mender-docs/blob/master/README-styleguide.markdown#use-second-person-you-when-speaking-to-or-about-the-reader) is what our styleguide says.

> When preparing client certificate for a device, the certificate key will be generated on a separate system (not on the device), and then provisioned into the device storage. This way we can keep records of the public key of the device and ensure sufficient entropy during key generation, so the resulting keys are secure random.

Author: Thanks, will use "you" form.

Then generate a certificate request from the newly generated private key:

```bash
openssl req -new -key device-private.key -out device-cert.req -config device-cert.conf
```

### Sign the client certificate

Now that we have both a CA certificate, and a certificate request for the device, we need to sign the latter with the former. This produces a signed certificate which the server will recognize when the client connects.

```bash
openssl x509 -req -CA ca-cert.pem -CAkey ca-private.key -CAcreateserial -in device-cert.req -out device-cert.pem -days $((365*10))
```

! The `-days` argument specifies how long the certificate is valid, and can be adjusted if needed. The example expression gives a certificate which is valid for approximately 10 years. Since the certificate will only be used by the server to authenticate devices, it is usually not desirable that it expires after a short time, since this requires certificate rotation on the devices. To manage compromised devices, it is often better to maintain a certificate blacklist on the server.

> The generation and signing of the client certificate needs to be repeated for each device, so these are natural steps to automate in your device provisioning workflow.

Reviewer 2: :+1:

> <!-- TODO: EVERYTHING FROM HERE AND DOWN TO "END_OF_SERVER_PART" NEEDS TO BE REWRITTEN. Originally copied from the Pre-Authorizing document. -->

Reviewer: Not sure I understand: can't we simply delete this section?

Author: Perhaps I should be more clear: What I meant that was a section needs to be written to set up the ambassador part with the certificate. I'll update the comment to reflect this.

<!-- END_OF_SERVER_PART -->

## Enable generated key and certificate in disk image

Now that we have generated a key and certificate for the device and signed the certificate, we need to copy them to our working disk image (it typically has the `.sdimg` suffix), and enable them in the Mender configuration.

### Copy key and certificate into disk image

Find the location of the [key and certificate we generated](#generate-certificates) and copy it into place on the data partition by running the following commands:

```bash
mender-artifact install -m 600 device-private.key mender-disk-image.sdimg:/data/mender/mender-cert-private.pem
mender-artifact install -m 644 device-cert.pem mender-disk-image.sdimg:/data/mender/mender-cert.pem
```

!!! The files we just added are per device, and therefore it is natural to automate this step in your device provisioning workflow.

### Set up Mender configuration for client certificate

First, copy the existing `mender.conf` out of the disk image, so that we can edit it.

```bash
mender-artifact cp mender-disk-image.sdimg:/etc/mender/mender.conf mender.conf
```

> Next, open `mender.conf` in any text editor, and add the following content:

Reviewer 2: ```suggestion
Next, open `mender.conf` in a text editor, and add the following content:
```

```json
  "HttpsClient": {
    "Certificate": "/data/mender/mender-cert.pem",
    "Key": "/data/mender/mender-cert.pem"
  }
```

> Make sure that the result is valid JSON, in particular that commas appear on every line except the last in a block. The snippet should be added inside the first set of curly braces in the file. For example, it might look like this in a typical `mender.conf` file:

Reviewer 2: The last comma thing is such a retarded part of JSON

Author: Omg, I know! This is like newline endings all over again. One design flaw, and it will haunt developers for decades.

```json
{
  "ServerURL": "https://hosted.mender.io/",
  "TenantToken": "TENANT_TOKEN",
  "HttpsClient": {
    "Certificate": "/data/mender/mender-cert.pem",
    "Key": "/data/mender/mender-cert.pem"
  }
}
```

Then copy the modified file back into the disk image:

```bash

> mender-artifact cp mender.conf mender-disk-image.sdimg:/etc/mender/mender.conf

Reviewer: One thing that is missing here is adding the CA cert as trusted, uploading it into `/etc/ssl/certs` and running `update-ca-certificates`.

Author: But does the client care whether it's trusted? If I've understood correctly, all trust checking happens on the server.

Author: @tranchitella: I did not get an answer to this. If you still think I should change something here, let me know, and I will follow up with a new PR.

```

!!! Since this change is the same on every device, it is natural to automate this as part of the build process for the disk image. See file installation instructions for [the Debian family](../../.04.System-updates-Debian-family/Customize-Mender#configuration-file) or [the Yocto Project](../../System-updates-Yocto-Project/Customize-Mender#configuration-file) for more information.

## Boot the device

Now provision the storage with this new disk image, just like you have done in the past. If you are using a SD card, insert it into your workstation and use a command similar to the following:

```bash
sudo dd if=<PATH-TO-IMAGE>.sdimg of=<DEVICE> bs=1M && sudo sync
```

Then insert the SD card back into your device and boot it.

> ## Verify the device is accepted

Reviewer 2: ```suggestion
## Verify that the device is accepted
```

> If everything went as intended, your device will get the `accepted` status in the Mender server. You can log in to the Mender UI to ensure your device is listed and reports inventory.

Reviewer 2: ```suggestion
If everything went as intended, your device shows up as `accepted` in the Mender server. You can log in to the Mender UI to ensure your device is listed and reports inventory.
```
